Networking v2: Add subnetpools support

ozerovandrei opened the pull request:
Add the support for Neutron subnetpools:
  - add the resourceNetworkingSubnetPoolV2
  - add the dataSourceNetworkingSubnetPoolV2
  - add acceptance tests and a documentation

For: #242

ozerovandrei commented:
recheck

jtopjian commented:
@ozerovandrei The build failure error is a little hard to find. You'll have to look in the raw output (`job-output.txt.gz` here: http://logs.openlabtesting.org/logs/43/243/d502eaf5e8d16dc2d2eef1f49b5d0be74ca510a7/check/terraform-provider-openstack-acceptance-test/3209b2f/)

```
* data.openstack_networking_subnetpool_v2.subnetpool_1: data.openstack_networking_subnetpool_v2.subnetpool_1: Your query returned more than one result. Please try a more specific search criteria
```

Searching for keyword "fail" will eventually lead to the error message.

If it is passing for you locally, then the issue is because the CI environment is creating a subnet that looks similar. It might take a few tries to get it correct. I had to do this with the Flavor acceptance tests a few weeks ago.

ozerovandrei commented:
@jtopjian yeah it's running smoothly in my environment.
It seems that it's failing on a datasource test query that searches for all subnetpools with `ip_version=4`.
I've just made less broad queries.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,245 @@
+				Computed: true,
+				ForceNew: true,
+			},
+			"created_at": &schema.Schema{
```
This might be a difficult one for the user to use. I recommend either removing it (I can't see a user querying a subnet pool by a specific date) or we'll need to ensure the correct date format is given.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,245 @@
+				Computed: true,
+				ForceNew: false,
+			},
+			"updated_at": &schema.Schema{
```
Same as `created_at`.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,245 @@
+	d.Set("project_id", subnetPool.ProjectID)
+	d.Set("created_at", subnetPool.CreatedAt)
+	d.Set("updated_at", subnetPool.UpdatedAt)
+	d.Set("prefixes", subnetPool.Prefixes)
```
When setting a non-scalar attribute, we now need to check for an error:

```go
if err := d.set("prefixes", subnetPool.Prefixes); err != nil {
  log.Printf("[WARN] unable to set prefixes: %s", err)
}
```

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2_test.go:
```diff
@@ -0,0 +1,123 @@
+	name = "subnetpool_1"
+	description = "terraform subnetpool acceptance test"
+
+	prefixes = ["10.10.0.0/16", "10.11.11.0/24"]
```
Thank you for specifically testing this to make sure the type cast works.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+				ForceNew: true,
+				Computed: true,
+			},
+			"created_at": &schema.Schema{
```
Since  you can't set `created_at`, this should be a "read-only" / computed-only attribute. You can do this by setting it by removing the `Optional` field in the schema.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+				ForceNew: false,
+				Computed: true,
+			},
+			"updated_at": &schema.Schema{
```
Same as `created_at`.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+				Optional: true,
+				ForceNew: false,
+			},
+			"revision_number": &schema.Schema{
```
This is another "read-only" / computed-only attribute.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
```
We should check for an error here.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+	d.Set("project_id", s.ProjectID)
+	d.Set("created_at", s.CreatedAt)
+	d.Set("updated_at", s.UpdatedAt)
+	d.Set("prefixes", s.Prefixes)
```
The same error check needs done here as in the data source.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,245 @@
+				Computed: true,
+				ForceNew: false,
+			},
+			"tenant_id": &schema.Schema{
```
Although List can probably accept both `tenant_id` and `project_id`, let's standardize on just `project_id`, so let's remove this.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+				Optional: true,
+				ForceNew: false,
+			},
+			"tenant_id": &schema.Schema{
```
Same as the data source, let's remove this and keep `project_id`.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,362 @@
+		return fmt.Errorf("Error deleting OpenStack Neutron Subnetpool: %s", err)
+	}
+
+	d.SetId("")
```
Small nit: this is no longer needed in the `Delete` calls. Terraform automatically does it.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,245 @@
+	}
+
+	if v, ok := d.GetOk("prefixes"); ok {
+		listOpts.Prefixes = v.([]string)
```
To protect against a strange type cast error, I wonder if we should call `resourceSubnetPoolPrefixesV2` here.

ozerovandrei commented:
@jtopjian thank you for a review.
I've never contributed to terraform or it's providers so these comments are really helpful.

ozerovandrei commented:
@jtopjian I rechecked list query with `prefixes`. It seems that it's not working. At least in my environment. Neutron just throws 500 internal server error and writes to log something like that:
```
File ".../neutron/objects/subnetpool.py", line 70, in modify_fields_to_db
ERROR neutron.api.v2.resource     for prefix in result['prefixes']
ERROR neutron.api.v2.resource KeyError: 'id'
```
So I removed `prefixes` from the list options.

I added `project_id` to update opts and fixed all issues that you commented.

jtopjian commented:
@ozerovandrei 

> I rechecked list query with prefixes. It seems that it's not working.

That's strange.

Can you try testing with both `TF_LOG=DEBUG` and `OS_DEBUG=1` set? That will display the API requests and responses. I wonder if the `prefixes` field isn't being formatted correctly. 

There will be _a lot_ of output, so it's best to pipe all output (stderr and stdout) to a file for review.

ozerovandrei commented:
@jtopjian I got the same error with a simple curl query to list subnetpools filtered by prefixes in a `stable/newton` release.
I installed a simple devstack with `stable/queens` release and right now I'm getting a different error if I'm trying to make a query to list subnetpools filtered by prefixes.:
```
{"NotImplementedError": {"message": "in_() not yet supported for relationships.  For a simple many-to-one, use in_() against the set of foreign key values.", "type": "NotImplementedError", "detail": ""}}
```.

I guess that I need to remove `prefixes` from `subnetpools.ListOpts` in the Gophercloud then. Sorry, I've never tested query with that option after I implemented `List()` for subnetpools 😞 .

jtopjian commented:
> I guess that I need to remove prefixes from subnetpools.ListOpts in the Gophercloud then. Sorry, I've never tested query with that option after I implemented List() for subnetpools 😞 .

It's no problem at all. There was probably something in the code that made it look as though a prefix query was possible or I would have never merged the code 😄 No harm done.

jtopjian commented:
@ozerovandrei I've submitted https://github.com/gophercloud/gophercloud/pull/789 which addresses the changes needed :)

ozerovandrei commented:
@jtopjian that's great. Thank you!

ozerovandrei commented:
@jtopjian do I need to change something else? I've removed `created_at`, `updated_at` and `prefixes` from requests.

jtopjian commented:
@ozerovandrei https://github.com/gophercloud/gophercloud/pull/789 was merged. If you haven't, can you do another vendor import just so the latest Gophercloud subnetpool package is bundled?

ozerovandrei commented:
@jtopjian I haven't noticed that `` and `` became `*bool` types.
How to better deal with that? Is something like that is a good choice (saw that in `subnets`):
```
data_source_openstack_networking_subnetpool_v2:
...
if v, ok := d.GetOk("shared"); ok {
    shared := true
    listOpts.Shared = &shared
}

if v, ok := d.GetOk("not_shared"); ok {
    shared := false
    listOpts.Shared = &shared
}

if v, ok := d.GetOk("default"); ok {
    isDefault := false
    listOpts.IsDefault = &isDefault
}

if v, ok := d.GetOk("not_default"); ok {
    isDefault := true
    listOpts.IsDefault = &isDefault
}
...
```
(Maybe `private` instead of `not_shared` is better).

jtopjian commented:
@ozerovandrei What's wrong with just having a single option of `shared` and passing the value (either `true` or `false`) to `ListOpts`?

ozerovandrei commented:
@jtopjian nothig wrong. It was an error assumption. All fixed 😄

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+			},
+			"created_at": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
```
This line needs to be removed to prevent the user from querying by `created_at`.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+			},
+			"updated_at": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
```
Same as `created_at`.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+			},
+			"prefixes": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
```
This needs removed, too.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+		listOpts.Description = v.(string)
+	}
+
+	if v, ok := d.GetOk("default"); ok {
```
Should this be `is_default`?

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+			},
+			"revision_number": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
```
I'm not getting an error when querying with `revision_number`, but not getting any results, either. It's probably best to remove `Optional` and make it a computed-only field.

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+		listOpts.IsDefault = v.(*bool)
+	}
+
+	if v, ok := d.GetOk("revision_number"); ok {
```
This will need removed

jtopjian commented on openstack/data_source_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,225 @@
+	}
+
+	if v, ok := d.GetOk("shared"); ok {
+		listOpts.Shared = v.(*bool)
```
Does this work? If so, that's great. But if not, you'll have to do:

```go
shared := v.(bool)
listOpts.Shared = &shared
```

ozerovandrei replied:
Sorry for the bad testing. Again. Yeah, it's not working, will fix that. (Same in "is_default").

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2.go:
```diff
@@ -0,0 +1,363 @@
+	}
+
+	if d.HasChange("is_default") {
+		isDefaultRaw := d.Get("is_default").(string)
```
You should be able to just do

```go
isDefault := d.Get("is_default").(bool)
```

There's really no need to do `GetOk` here, so `Get` is fine.

ozerovandrei replied:
I don't get it. There is no `GetOk` in resource_openstack_networking_subnetpool_v2.go.

jtopjian replied:
Sorry - my comment was confusing. I was confirming that using `Get` is fine.

But the important change that needs done here is to cast to a `bool` rather than a `string`.

jtopjian commented on website/docs/d/networking_subnetpool_v2.html.markdown:
```diff
@@ -0,0 +1,78 @@
+
+* `is_default` - (Optional) Whether the subnetpool is default subnetpool or not.
+
+* `revision_number` - The revision number of the subnetpool.
```
This will need removed, too

jtopjian commented on website/docs/d/networking_subnetpool_v2.html.markdown:
```diff
@@ -0,0 +1,78 @@
+* `shared` - See Argument Reference above.
+* `description` - See Argument Reference above.
+* `is_default` - See Argument Reference above.
+* `revision_number` - See Argument Reference above.
```
And this will need updated with `The revision number of the subnetpool.`

jtopjian commented on website/docs/r/networking_subnetpool_v2.html.markdown:
````diff
@@ -0,0 +1,106 @@
+```hcl
+resource "openstack_networking_subnetpool_v2" "subnetpool_1" {
+  name = "my_ipv6_pool"
+  prefixes = ["fdf7:b13d:dead:beef::/64", "fd65:86cc:a334:39b7::/64"]
````
I think `ip_version` defaults to "4", so in order for the above example to be valid, you'll have to add `ip_version = 6` here.

jtopjian commented on openstack/resource_openstack_networking_subnetpool_v2_test.go:
```diff
@@ -0,0 +1,151 @@
+}
+`
+
+const testAccNetworkingV2SubnetPoolPrefixLenghts = `
```
This is a really good test to have - nice. There's a typo in the name, though: `Lengths`.

Maybe a third step should be added which updates some of the fields? That way it tests the update functionality.

jtopjian commented:
@ozerovandrei This looks good! Thank you for your patience and work with this. These are really nice additions to Terraform :)

ozerovandrei commented:
@jtopjian thank you for such a thorough review 😄

lunderhage commented:
Would it be possible to get some example about how to create a subnet using a subnetpool? The documentation does not say much for now.

jtopjian commented:
@lunderhage I apologize for the delay -- I lost track of this.

We've just merged a doc fix for this here: https://github.com/terraform-providers/terraform-provider-openstack/pull/312

It'll appear on the site next week.
